drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,299 @@
+#define THREAD_PRIORITY   K_PRIO_PREEMPT(CONFIG_AT_CMD_THREAD_PRIO)
+
+#define AT_CMD_OK_STR    "OK"
+#define AT_CMD_ERROR_STR "ERROR"
+#define AT_CMD_CMS_STR   "+CMS:"
+#define AT_CMD_CME_STR   "+CME:"
+
+static K_THREAD_STACK_DEFINE(socket_thread_stack, CONFIG_AT_CMD_THREAD_STACK_SIZE);
+static K_MUTEX_DEFINE(cmd_pending);
+
+K_MSGQ_DEFINE(return_code_msq, sizeof(enum at_cmd_rc), 2, 4);
+
+static int              common_socket_fd;
+static int              last_error;
+static char             *response_buf;
+static u32_t            response_buf_len;
+
+static struct k_thread  socket_thread;
+static at_cmd_handler_t notification_handler;
+static at_cmd_handler_t current_cmd_handler;
+
+struct callback_work_item {
+	struct k_work    work;
+	char             data[RESPONSE_MAX_LEN];
+	at_cmd_handler_t callback;
+};
+
+static int open_socket(void)
+{
+	common_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
+
+	if (common_socket_fd == -1) {
+		LOG_ERR("AT socket creation failed");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static enum at_cmd_rc get_return_code(char *buf)
+{
+	enum at_cmd_rc return_code = AT_CMD_NOTIFICATION;
+	char *tmpstr = NULL;
+
+	do {
+		tmpstr = strstr(buf, AT_CMD_OK_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_OK;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_ERROR_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_ERROR_CMS) ||
+	    (return_code == AT_CMD_ERROR_CME)) {
+		last_error = atoi(&buf[ARRAY_SIZE(AT_CMD_CMS_STR) - 1]);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	if ((data != NULL) && (data->callback != NULL)) {
+		data->callback(data->data);
+	}
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("AT socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+		write_with_response = false;
+
+		if (bytes_read < 0) {
+			LOG_ERR("AT socket recv failed with err %d",
+				bytes_read);
+
+			if ((close(common_socket_fd) == 0) &&
+			    (open_socket() == 0)) {
+				LOG_INF("AT socket recovered");
+
+				if (cmd_pending.lock_count > 0) {
+					return_code = AT_CMD_ERROR;
+					goto next;
+				} else {
+					continue;
+				}
+			}
+
+			LOG_ERR("Unrecoverable reception error (err: %d), "
+				"thread killed", errno);
+			return;
```

Comment:
Should you explicitly close the socket here?

Reply:
ok